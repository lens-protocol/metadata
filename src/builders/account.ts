import { v4 } from 'uuid';

import {
  type AccountMetadata,
  type AccountMetadataDetails,
  AccountMetadataSchema,
  AccountMetadataSchemaId,
} from '../account';
import type { Prettify } from '../utils.js';
import { evaluate } from './ValidationError.js';
import type { RecursiveUnbrand } from './utils.js';

/**
 * @private
 * @privateRemarks MUST stay very @private to produce usable docs
 */
type AccountDetails = Prettify<RecursiveUnbrand<Omit<AccountMetadataDetails, 'id'>>>;
/**
 * All {@link AccountMetadataDetails} fields with:
 */
export type AccountOptions = AccountDetails & {
  /**
   * A unique identifier that in storages like IPFS ensures the uniqueness of the metadata URI.
   *
   * @defaultValue a UUID
   */
  id?: string;
};
/**
 * Creates a valid AccountMetadata.
 *
 * @category Compose
 * @param input - Use your IDE suggestions for an enhanced development experience
 *
 * @example
 * ```ts
 * const metadata = account({
 *   name: 'John Doe',
 *   bio: `
 *   Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.
 *
 *   - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
 *   - Donec a diam lectus.
 *   `,
 * });
 * ```
 *
 * @example
 * With attributes:
 * ```ts
 * const metadata = account({
 *   name: 'John Doe',
 *   picture: 'https://example.com/picture.png',
 *   attributes: [
 *     {
 *       key: 'twitter',
 *       type: MetadataAttributeType.STRING,
 *       value: 'https://x.com/johndoe',
 *     },
 *     {
 *       key: 'dob',
 *       type: MetadataAttributeType.DATE,
 *       value: '1990-01-01T00:00:00Z',
 *     },
 *     {
 *       key: 'enabled',
 *       type: MetadataAttributeType.BOOLEAN,
 *       value: 'true',
 *     },
 *     {
 *       key: 'height',
 *       type: MetadataAttributeType.NUMBER,
 *       value: '1.8',
 *     },
 *     {
 *       key: 'settings',
 *       type: MetadataAttributeType.JSON,
 *       value: '{"theme": "dark"}',
 *     },
 *   ],
 * });
 * ```
 */
export function account({ id = v4(), ...others }: AccountOptions): AccountMetadata {
  return evaluate(
    AccountMetadataSchema.safeParse({
      $schema: AccountMetadataSchemaId.LATEST,
      lens: {
        id,
        ...others,
      },
    }),
  );
}
