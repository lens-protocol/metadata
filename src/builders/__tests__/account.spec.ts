import { describe, expect, it } from 'vitest';

import { MetadataAttributeType } from '../../MetadataAttribute.js';
import { account } from '../account.js';

describe(`Given the "${account.name}" metadata builder`, () => {
  describe('when using it', () => {
    it('should return a valid AccountMetadata', () => {
      const metadata = account({
        name: 'John Doe',
        bio: `
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.

        - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        - Donec a diam lectus.
        `,
        picture: 'https://example.com/picture.png',
        coverPicture: 'https://example.com/cover.png',
        attributes: [
          {
            key: 'twitter',
            type: MetadataAttributeType.STRING,
            value: 'https://x.com/johndoe',
          },
          {
            key: 'dob',
            type: MetadataAttributeType.DATE,
            value: '1990-01-01T00:00:00Z',
          },
          {
            key: 'enabled',
            type: MetadataAttributeType.BOOLEAN,
            value: 'true',
          },
          {
            key: 'height',
            type: MetadataAttributeType.NUMBER,
            value: '1.8',
          },
          {
            key: 'settings',
            type: MetadataAttributeType.JSON,
            value: '{"theme": "dark"}',
          },
        ],
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });
});
