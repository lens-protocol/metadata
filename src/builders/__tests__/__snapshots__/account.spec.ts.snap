// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given the "account" metadata builder > when using it > should return a valid AccountMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/account/1.0.0.json",
  "lens": {
    "attributes": [
      {
        "key": "twitter",
        "type": "String",
        "value": "https://x.com/johndoe",
      },
      {
        "key": "dob",
        "type": "Date",
        "value": "1990-01-01T00:00:00Z",
      },
      {
        "key": "enabled",
        "type": "Boolean",
        "value": "true",
      },
      {
        "key": "height",
        "type": "Number",
        "value": "1.8",
      },
      {
        "key": "settings",
        "type": "JSON",
        "value": "{"theme": "dark"}",
      },
    ],
    "bio": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.

        - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        - Donec a diam lectus.",
    "coverPicture": "https://example.com/cover.png",
    "id": Any<String>,
    "name": "John Doe",
    "picture": "https://example.com/picture.png",
  },
}
`;
